File: Data-Structures/Linked-List/Singly_Linked_List/Detect_Cycle/detect_cycle.cs
using System;
namespace DSA
{
    public class Node
    {
        public int Value { get; set; }
        public Node Next;
        public Node(int val)
        {
            this.Value = val;
        }
    }

    class ClosedLoop
    {
        public static bool DetectLoop(Node _head)
        {
            Node slow = _head, fast = _head.Next.Next;
            while (slow != null && fast != null && fast.Next != null)
            {
                if (slow == fast)
                {
                    Console.WriteLine("Loop detected at: " + slow.Value);
                    return true;
                }
                slow = slow.Next; //1 step increment
                fast = fast.Next.Next; //2 step increment
            }
            return false;
        }

        static void Main(string[] args)
        {
            Node tail = new Node(0);
            Node _actualHead = tail;

            Node intersection = null;

            //Creating linked list by adding nodes to the end of the list.
            for (int i = 1; i < 10; i++)
            {
                Node n = new Node(i);
                tail.Next = n;
                tail = n;
                //Taking the loop intersection. Comment below 'if' to avoid loop.
                if (i == 4)
                    intersection = tail;
            }

            //Completing the loop
            tail.Next = intersection;

            if (!DetectLoop(_actualHead))
                Console.WriteLine("No loop detected");
        }
    }
}

File: Data-Structures/Linked_List/Singly_Linked_List/Detect_Cycle/detect_cycle.go
package main

import "fmt"

type Node struct {
	Value int
	Next *Node
}

func NewNode(value int) *Node {
	return &Node{
		Value: value,
		Next:  nil,
	}
}

func DetectLoop(head *Node) bool {
	slow := head
	fast := head
	for slow != nil && fast != nil && fast.Next != nil{
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast && slow != nil{
			fmt.Println("Loop detected at ",slow.Value)
			return true
		}
	}
	return false
}

func main()  {
	tail := NewNode(0)
	_actualHead := tail
	var intersection *Node
	// creating linked list
	for i := 1; i < 10; i++ {
		n := NewNode(i)
		tail.Next = n
		tail = n

		if i == 5 {
			intersection = tail // marking the loop node
		}
	}
	// completing the loop
	tail.Next = intersection
	if !DetectLoop(_actualHead) {
		fmt.Println("No loop detected")
	}
}




